exporter/v1_compatibily_test.go

```diff
@@ -0,0 +1,89 @@
+package exporter
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+func TestWalkTo(t *testing.T) {
+	m := bson.M{
+		"serverStatus": bson.M{
+			"locks": bson.M{
+				"ParallelBatchWriterMode": bson.M{
+					"acquireCount": bson.M{
+						"r": float64(1.23),
+					},
+				},
+			},
+		},
+	}
+
+	testCases := []struct {
+		path []string
+		want interface{}
+	}{
+		{
+			path: []string{"serverStatus", "locks", "ParallelBatchWriterMode", "acquireCount", "r"},
+			want: float64(1.23),
+		},
+		{
+			path: []string{"serverStatus", "locks", "ParallelBatchWriterMode", "acquireCount", "r", "w"},
+			want: nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		assert.Equal(t, walkTo(m, tc.path), tc.want)
+	}
+}
+
+func TestMakeLockMetric(t *testing.T) {
+	m := bson.M{
+		"serverStatus": bson.M{
+			"locks": bson.M{
+				"ParallelBatchWriterMode": bson.M{
+					"acquireCount": bson.M{
+						"r": float64(1.23),
+					},
+				},
+			},
+		},
+	}
+
+	lm := lockMetric{
+		name:   "mongodb_ss_locks_acquireCount",
+		path:   strings.Split("serverStatus_locks_ParallelBatchWriterMode_acquireCount_r", "_"),
+		labels: map[string]string{"resource": "ParallelBatchWriterMode", "lock_mode": "r"},
+	}
+
+	want := `Desc{fqName: "mongodb_ss_locks_acquireCount", ` +
+		`help: "mongodb_ss_locks_acquireCount", ` +
+		`constLabels: {}, variableLabels: [resource lock_mode]}`
+
+	p, err := makeLockMetric(m, lm)
+	assert.NoError(t, err)
+	assert.Equal(t, want, p.Desc().String())
+}
+
+func TestAddLocksMetrics(t *testing.T) {
+	buf, err := ioutil.ReadFile(filepath.Join("testdata/", "locks.json"))
+	assert.NoError(t, err)
+
+	var m bson.M
+	err = json.Unmarshal(buf, &m)
+	assert.NoError(t, err)
+
+	var metrics []prometheus.Metric
+	metrics = addLocksMetrics(m, metrics)
+
+	// labels are not being returned in the same order, so it is hard to test the real metric descriptions
```

Comment:
It is not about sorting metrics.
Metric fields are unexported so we cannot compare them. If we use String() since metric labels are stored as maps, the order is not always the same.

Reply:
why can't we sort results of calling String()?

Reply:
We can sort results of String() and then use method `Equal` to compare with expected results.